logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/MLSConversationsVerificationStatusesHandler.kt

```diff
@@ -59,32 +79,108 @@ internal class MLSConversationsVerificationStatusesHandlerImpl(
+                        } else {
+                            conversationRepository.getConversationDetailsByMLSGroupId(groupId).map {
+                                VerificationStatusData(
+                                    conversationId = it.conversation.id,
+                                    currentStatusInDatabase = it.conversation.mlsVerificationStatus,
+                                    newStatus =
+                                    ccGroupStatus
+                                )
+                            }
+                        }
+                    }.onSuccess {
+                        updateStatusAndNotifyUserIfNeeded(it)
+                    }
             }
-            .onlyRight()
-            .collect { (groupId, newStatus) ->
-                conversationRepository.getConversationDetailsByMLSGroupId(groupId)
-                    .map { conversation -> updateStatusAndNotifyUserIfNeeded(newStatus, conversation) }
+    }
+
+    private suspend fun verifyUsersStatus(groupId: GroupID): Either<CoreFailure, VerificationStatusData> =
+        conversationRepository.getGroupStatusMembersNamesAndHandles(groupId)
+            .flatMap { epochChangesData ->
+                mlsConversationRepository.getMembersIdentities(
+                    epochChangesData.conversationId.toModel(),
+                    epochChangesData.members.keys.map { it.toModel() })
+                    .flatMap { ccIdentities ->
+                        updateKnownUsersIfNeeded(epochChangesData, ccIdentities, groupId)
+                    }
+            }.map { (dbData, ccIdentity) ->
+                var newStatus: VerificationStatus = VerificationStatus.VERIFIED
+                // check that all identities are valid and name and handle are matching
+                for ((userId, wireIdentity) in ccIdentity) {
+                    val memberInfoInDatabase = dbData.members[userId.toDao()]
+                    val isUserVerified = wireIdentity.firstOrNull {
+                        it.status != CryptoCertificateStatus.VALID ||
+                                it.displayName != memberInfoInDatabase?.name ||
+                                it.handle != memberInfoInDatabase.handle
+                    } == null
+                    if (!isUserVerified) {
+                        newStatus = VerificationStatus.NOT_VERIFIED
+                        break
+                    }
+                }
+                VerificationStatusData(
+                    conversationId = dbData.conversationId.toModel(),
+                    currentStatusInDatabase = dbData.mlsVerificationStatus.toModel(),
+                    newStatus = newStatus
+                )
             }
+
+    private suspend fun updateKnownUsersIfNeeded(
+        epochChangesData: EpochChangesDataEntity,
+        ccIdentities: Map<UserId, List<WireIdentity>>,
+        groupId: GroupID
+    ): Either<CoreFailure, Pair<EpochChangesDataEntity, Map<UserId, List<WireIdentity>>>> {
```

Comment:
A bit hard to maintain when dealing with results of such kind.

It would be easier to know what is going on by having a simple data class instead, then dealing with a List, within a Map, within a Pair, within an Either.

Maybe `Map<UserId, List<WireIdentity>` could be made into a data class.

The Pair maybe as well?